lib/ansible/executor/module_common.py

```diff
@@ -1334,13 +1350,21 @@ def get_action_args_with_defaults(action, args, defaults, templar):
         module_defaults = templar.template(module_defaults)
 
         # deal with configured group defaults first
-        if action in C.config.module_defaults_groups:
-            for group in C.config.module_defaults_groups.get(action, []):
-                tmp_args.update((module_defaults.get('group/{0}'.format(group)) or {}).copy())
+        for group_name in collection_groups:
+            if 'group/%s' % group_name not in module_defaults:
+                continue
+
+            for collection_name in collection_groups[group_name]:
```

Comment:
Doing this inline is OK for beta- it makes the feature function, but it's a performance problem. Post-beta, we'll want to move most of this to an intermediate lazy indexing process during playbook parsing when a module_defaults clause references a group, where we'd build that global "sieve" I was talking about for O:1 lookup. At that point, we can probably get rid of the transformation in the collection metadata canonicalization phase entirely and just store it plain.

Reply:
Sounds good, will plan on a follow up PR.